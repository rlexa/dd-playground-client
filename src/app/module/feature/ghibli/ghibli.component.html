<div ddFlexbox="hyscsw">
  <mat-card appearance="outlined" class="childhv">
    <mat-card-title>Queries</mat-card-title>
    <mat-card-content>
      <div ddFlexbox="vys">
        <mat-list class="mat-list-auto-height">
          <h3 mat-subheader>API</h3>
          @for (type of TYPES_LISTS; track $index) {
            <mat-list-item>
              <button mat-raised-button (click)="toAnyData(typeToList$[type])">{{ type }}</button>
            </mat-list-item>
          }
        </mat-list>
      </div>
    </mat-card-content>
  </mat-card>

  <mat-card appearance="outlined" *ngIf="tableAllData$ | async as tableAllData" class="childhv">
    <mat-card-title>List</mat-card-title>
    <mat-card-content>
      <app-loading-bar [show]="(busyCount$ | async) > 0" />
      <app-simple-table
        [style.maxWidth]="'100rem'"
        [data]="tableAllData"
        [isClickable]="clickableValue"
        (clicked)="onClickedValue($event.key, $event.value)"
      />
    </mat-card-content>
  </mat-card>

  <mat-card appearance="outlined" *ngIf="anyData$ | async as anyData" class="childhv">
    <mat-card-title>Details</mat-card-title>
    <mat-card-content class="scroll-y-auto" [style.maxHeight]="'60rem'">
      <app-loading-bar [show]="(busyCount$ | async) > 0" />
      <app-simple-view
        [data]="anyData"
        [isWidthConstrained]="false"
        [isDense]="true"
        [isClickable]="clickableValue"
        (clicked)="onClickedValue($event.key, $event.value)"
      />
    </mat-card-content>
  </mat-card>
</div>
