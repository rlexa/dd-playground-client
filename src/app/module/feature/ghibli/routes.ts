import {Route} from '@angular/router';
import {RouteRoot, RouteWild} from 'src/app/routing';

export default [
  {path: RouteRoot, loadComponent: () => import('./ghibli.component').then((ii) => ii.GhibliComponent)},
  {path: RouteWild, redirectTo: RouteRoot},
] as Route[];
