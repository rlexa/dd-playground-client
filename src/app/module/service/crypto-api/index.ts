export * from './crypto-api.module';
export * from './crypto-api.service';
