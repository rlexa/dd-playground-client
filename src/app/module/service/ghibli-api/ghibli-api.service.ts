import {HttpClient} from '@angular/common/http';
import {Injectable} from '@angular/core';
import {cleanParams, getJsonHeaders} from 'src/app/util';
import {map} from 'rxjs/operators';

const API = 'https://ghibliapi.herokuapp.com';

export const GHIBLI_TYPE_FILM = '/films/';
export const GHIBLI_TYPE_LOCATION = '/locations/';
export const GHIBLI_TYPE_PEOPLE = '/people/';
export const GHIBLI_TYPE_SPECIES = '/species/';
export const GHIBLI_TYPE_VEHICLES = '/vehicles/';
export const GHIBLI_TYPES = [GHIBLI_TYPE_FILM, GHIBLI_TYPE_LOCATION, GHIBLI_TYPE_PEOPLE, GHIBLI_TYPE_SPECIES, GHIBLI_TYPE_VEHICLES];

export interface GhibliEntity {
  id: string;
  ghibliType: string;
}

const removeApiPrefix = (val: unknown) => (typeof val === 'string' && val.startsWith(API) ? val.substr(API.length) : val);

const sanitizeApiPrefix = (item: GhibliEntity) =>
  Object.entries(item || {}).forEach(([key, val]) =>
    typeof val === 'string'
      ? (item[key] = removeApiPrefix(val))
      : Array.isArray(val)
      ? (item[key] = val.map(removeApiPrefix))
      : typeof val === 'object'
      ? (item[key] = sanitizeApiPrefix(val))
      : {},
  );

const mapGhibliEntity = (from: any) => {
  const url = Array.isArray(from.url) ? from.url[0] : from.url;
  const ret: GhibliEntity = {...(from || {})};
  ret.ghibliType = typeof url === 'string' ? GHIBLI_TYPES.find(_ => url.includes(_)) : '';
  ret.id = ret.ghibliType + ret.id;

  sanitizeApiPrefix(ret);

  return ret;
};

@Injectable({providedIn: 'root'})
export class GhibliApiService {
  constructor(private readonly http: HttpClient) {}

  private readonly API = API;

  film$ = (id: string) => this.httpGetRest$({url: `/films/${id}`}).pipe(map(mapGhibliEntity));
  films$ = () => this.httpGetRest$({url: `/films`}).pipe(map((_: object[]) => _.map(mapGhibliEntity)));
  location$ = (id: string) => this.httpGetRest$({url: `/locations/${id}`}).pipe(map(mapGhibliEntity));
  locations$ = () => this.httpGetRest$({url: `/locations`}).pipe(map((_: object[]) => _.map(mapGhibliEntity)));
  person$ = (id: string) => this.httpGetRest$({url: `/people/${id}`}).pipe(map(mapGhibliEntity));
  persons$ = () => this.httpGetRest$({url: `/people`}).pipe(map((_: object[]) => _.map(mapGhibliEntity)));
  species$ = (id: string) => this.httpGetRest$({url: `/species/${id}`}).pipe(map(mapGhibliEntity));
  speciess$ = () => this.httpGetRest$({url: `/species`}).pipe(map((_: object[]) => _.map(mapGhibliEntity)));
  vehicle$ = (id: string) => this.httpGetRest$({url: `/vehicles/${id}`}).pipe(map(mapGhibliEntity));
  vehicles$ = () => this.httpGetRest$({url: `/vehicles`}).pipe(map((_: object[]) => _.map(mapGhibliEntity)));

  private httpGetRest$ = (val: {api?: string; url: string; params?: any}) =>
    this.http.get((val.api || this.API) + val.url, {headers: getJsonHeaders(), params: cleanParams(val.params)});
}
