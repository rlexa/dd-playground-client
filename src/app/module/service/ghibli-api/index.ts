export * from './ghibli-api.service';
