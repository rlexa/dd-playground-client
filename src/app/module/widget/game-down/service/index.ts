export * from './game-down.service';
