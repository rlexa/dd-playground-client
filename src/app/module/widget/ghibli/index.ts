export * from './ghibli.component';
export * from './ghibli.module';
