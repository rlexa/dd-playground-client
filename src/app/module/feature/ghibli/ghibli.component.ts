import {CommonModule} from '@angular/common';
import {ChangeDetectionStrategy, Component, OnDestroy, inject} from '@angular/core';
import {MatButtonModule} from '@angular/material/button';
import {MatCardModule} from '@angular/material/card';
import {MatListModule} from '@angular/material/list';
import {DoneSubject, RxCleanup, rxNext_} from 'dd-rxjs';
import {BehaviorSubject, Observable, Subject, of} from 'rxjs';
import {filter, finalize, map, switchMap, tap} from 'rxjs/operators';
import {GHIBLI_TYPES, GhibliApiService, GhibliEntity, GhibliType} from 'src/app/module/service/ghibli-api';
import {FlexboxDirective} from '../../directive/flexbox';
import {LoadingBarComponent} from '../../widget/loading-bar';
import {SimpleTableComponent} from '../../widget/simple-table';
import {SimpleViewComponent} from '../../widget/simple-view';

@Component({
  selector: 'app-ghibli',
  templateUrl: './ghibli.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
  standalone: true,
  imports: [
    CommonModule,
    MatButtonModule,
    MatCardModule,
    MatListModule,
    FlexboxDirective,
    LoadingBarComponent,
    SimpleTableComponent,
    SimpleViewComponent,
  ],
})
export class GhibliComponent implements OnDestroy {
  private readonly api = inject(GhibliApiService);

  private readonly typeToDetails$: Record<GhibliType, (id: string) => Observable<GhibliEntity>> = {
    [GhibliType.Location]: this.api.location$,
    [GhibliType.Movie]: this.api.film$,
    [GhibliType.Person]: this.api.person$,
    [GhibliType.Species]: this.api.species$,
    [GhibliType.Vehicles]: this.api.vehicle$,
  };

  @RxCleanup() private readonly done$ = new DoneSubject();

  readonly typeToList$: Record<GhibliType, () => Observable<GhibliEntity[]>> = {
    [GhibliType.Location]: this.api.locations$,
    [GhibliType.Movie]: this.api.films$,
    [GhibliType.Person]: this.api.persons$,
    [GhibliType.Species]: this.api.speciess$,
    [GhibliType.Vehicles]: this.api.vehicles$,
  };
  readonly TYPES_LISTS = Object.keys(this.typeToList$);

  @RxCleanup() readonly anyData$ = new Subject();
  @RxCleanup() readonly busyCount$ = new BehaviorSubject(0);

  readonly tableAllData$ = this.anyData$.pipe(map((_) => (Array.isArray(_) ? _ : null)));

  destroy() {}
  ngOnDestroy() {
    this.destroy();
  }

  clickableValue = (key: string, value: any) => typeof value === 'string' && GHIBLI_TYPES.findIndex((_) => value.includes(_)) >= 0;

  onClickedValue = (key: string, value: any) => {
    if (typeof value === 'string') {
      const ghibliType = GHIBLI_TYPES.find((_) => value.includes(_));
      if (ghibliType) {
        const id = value.substr(ghibliType.length);
        const getter = (id ? this.typeToDetails$ : this.typeToList$)[ghibliType];
        this.toAnyData(getter, id);
      }
    }
  };

  toAnyData = <T>(switchTo: (val: T) => Observable<any>, param?: T) =>
    of(param)
      .pipe(
        filter((_) => !!switchTo),
        tap(() => this.busyCount$.next(this.busyCount$.value + 1)),
        switchMap(switchTo),
        finalize(() => this.busyCount$.next(this.busyCount$.value - 1)),
      )
      .subscribe(rxNext_(this.anyData$), rxNext_(this.anyData$));
}
