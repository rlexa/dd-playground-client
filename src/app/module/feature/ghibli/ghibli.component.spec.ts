import {ComponentFixture} from '@angular/core/testing';
import {MockBuilder, MockRender} from 'ng-mocks';
import {mockAll} from 'src/app/test';
import {GhibliApiService} from '../../service/ghibli-api';
import {GhibliComponent} from './ghibli.component';

describe('GhibliComponent', () => {
  let fixture: ComponentFixture<GhibliComponent>;

  beforeEach(() => MockBuilder(GhibliComponent).provide(mockAll(GhibliApiService)));

  beforeEach(() => {
    fixture = MockRender(GhibliComponent);
    fixture.detectChanges();
  });

  test('renders', () => expect(fixture).toMatchSnapshot());
});
