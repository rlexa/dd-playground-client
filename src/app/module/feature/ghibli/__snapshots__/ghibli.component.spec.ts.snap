// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GhibliComponent renders 1`] = `
<mock-render>
  <app-ghibli>
    <div
      ddflexbox="hyscsw"
      ng-reflect-dd-flexbox="hyscsw"
    >
      <mat-card
        appearance="outlined"
        class="childhv"
        ng-reflect-appearance="outlined"
      >
        <mat-card-title>
          Queries
        </mat-card-title>
        <mat-card-content>
          <div
            ddflexbox="vys"
            ng-reflect-dd-flexbox="vys"
          >
            <mat-list
              class="mat-list-auto-height"
            >
              <h3
                mat-subheader=""
              >
                API
              </h3>
              <mat-list-item>
                <button
                  mat-raised-button=""
                >
                  /locations/
                </button>
              </mat-list-item>
              <mat-list-item>
                <button
                  mat-raised-button=""
                >
                  /films/
                </button>
              </mat-list-item>
              <mat-list-item>
                <button
                  mat-raised-button=""
                >
                  /people/
                </button>
              </mat-list-item>
              <mat-list-item>
                <button
                  mat-raised-button=""
                >
                  /species/
                </button>
              </mat-list-item>
              <mat-list-item>
                <button
                  mat-raised-button=""
                >
                  /vehicles/
                </button>
              </mat-list-item>
            </mat-list>
          </div>
        </mat-card-content>
      </mat-card>
    </div>
  </app-ghibli>
</mock-render>
`;
